### ISSUE TITLE
Some packages installed with win_chocolatey (git/hg) are not available on command-line until log out/log in

### ISSUE THREAD

**Issue by chrisvire** [ansible/ansible#24333]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

win_chocolatey

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.3.0.0
  config file = /home/hubbard/src/ops-ansible-sites/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
pipelining = true
 
##### OS / ENVIRONMENT
Control Machine = Ubuntu 16.04
Managed Host = Windows 7 64-bit

##### SUMMARY
When using win_chocolatey to install some packages (git, hg), the environment is changed (going to advanced system settings -> environment, I can see that the PATH has been changed), but the commands cannot be executed in a command-line by the logged in user until the user logs out and logs in again.

##### STEPS TO REPRODUCE
```yaml
- name: "Packages: install version control packages"
  win_chocolatey: name={{item}}
  with_items:
    - git
    - hg
```
##### EXPECTED RESULTS
After executing the playbook, I expected the logged in user to be able to execute the commands.  

##### ACTUAL RESULTS
After executing the playbook, the commands cannot be executed from the logged in user in a a newly started cmd.exe window.  Running the path command doesn't display the updated PATH. 

If I start a cmd.exe with "Run As Administrator" and run the path command, then the correct PATH environment variable is displayed.

**Comment by ansibot** [comment]
cc @smadam813 @nitzmahone
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chrisvire** [comment]
This isn't a high priority since I can document this issue, but I thought it should be reported.

Is there a way to know in the Ansible script that it was actually installed and cause a reboot?  Maybe register an event and have a win_reboot entry with a when clause?

**Comment by chrisvire** [comment]
Anyone looked at this?  I also noticed that if I run a Powershell script to install fonts, the fonts are not seen by applications until the user logs out and logs in again.

**Comment by jhawkesworth** [comment]
I have a feeling this is just 'how windows does things' unfortunately.
Environment variables are usually regarded as immutable once a program has started.
There are exceptions and a mechanism was created whereby executables could listen for change events when environment variables have been modified, but it is rare for executables to implement this in my experience (I think I read somewhere that 'explorer' does listen).

Regarding what to do about it, there has been an 'ansible_reboot_pending' fact for a while now, so you could run the setup module again and then do a win_reboot with a when: clause to check if ansible_reboot_pending is true.

Hope this helps,

Jon

**Comment by jhawkesworth** [comment]
@chrisvire - there was a discussion on the #ansible-windows channel a few days ago and it was suggested that by running the 'setup' module to gather facts again, a new remote shell would be opened, which would make new environment variables available.

can you give it a try - I am hopeful it will fix the problem described for you?

needs_info

**Comment by dagwieers** [comment]
At [19:00 UTC](http://www.thetimezoneconverter.com/?t=19:00&tz=UTC) this ticket will be processed during [the first Windows Sprint](https://github.com/ansible/community/wiki/Windows:-sprints#2017-07-04-1900-utc-sprint-1), we would appreciate if you could join the Sprint to help out with fixing/closing this issue.

**Comment by jhawkesworth** [comment]
Used the following playbook to try and reproduce this, but found I was able to run git commands straight away after win_chocolatey had completed, even before running setup.
```yaml

- name: choco test env setup
  hosts: 10,
  tasks:
    - name: install git
      win_chocolatey:
        name: git

    - name: run git command
      win_shell: git status

    - name: gather facts again to open new shell
      setup:

    - name: run git command again
      win_shell: git status
```

results were:
```

No config file found; using defaults

PLAY [choco test env setup] *************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************************
ok: [TENSY.home]

TASK [install git] **********************************************************************************************************************
changed: [TENSY.home] => {"changed": true, "failed": false, "rc": 0}

TASK [run git command] ******************************************************************************************************************
changed: [TENSY.home] => {"changed": true, "cmd": "git status", "delta": "0:00:00.439772", "end": "2017-07-04 10:15:01.163393", "failed": false, "rc": 1, "start": "2017-07-04 10:15:00.723620", "stderr": "fatal: Not a git repository (or any of the parent directories): .git", "stderr_lines": ["fatal: Not a git repository (or any of the parent directories): .git"], "stdout": "", "stdout_lines": []}

TASK [gather facts again to open new shell] *********************************************************************************************
ok: [TENSY.home]

TASK [run git command again] ************************************************************************************************************
changed: [TENSY.home] => {"changed": true, "cmd": "git status", "delta": "0:00:00.466825", "end": "2017-07-04 10:15:10.138575", "failed": false, "rc": 1, "start": "2017-07-04 10:15:09.671749", "stderr": "fatal: Not a git repository (or any of the parent directories): .git", "stderr_lines": ["fatal: Not a git repository (or any of the parent directories): .git"], "stdout": "", "stdout_lines": []}

PLAY RECAP ******************************************************************************************************************************
TENSY.home                 : ok=5    changed=3    unreachable=0    failed=0

```

**Comment by jborean93** [comment]
For logged in users there is nothing we can really do unfortunately as that's the way Windows works. This is an issue as there is notify process that can be run that tells a logged in profile to reload the environment. I'm surprised that opening up a new cmd instance doesn't load the new environment variables so I'll have a look and see if I can reproduce it on my end.

For Ansible everytime we run a new task a new shell is created so it will inherit the latest environment variables like @jhawkesworth has shown.

**Comment by chrisvire** [comment]
@dagwieers, sorry I can't join the sprint.

@jhawkesworth, If I run a admin shell, then it works fine (programs are on the PATH).  The issue is with the logged in user.  This affects both PATH and fonts that are available.

I haven't tried this yet: "you could run the setup module again and then do a win_reboot with a when: clause to check if ansible_reboot_pending is true."  I now have other higher priority tasks that I need to complete before I go on vacation.  I will see if I can fit this in.

**Comment by ansibot** [comment]
cc @elventear @petemounce @trondhindenes
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jborean93** [comment]
I am going to close this as there really isn't anything we can do about it. Whenever a new shell is created the new PATH is loaded. Unfortunately Windows does not expose any knobs (that I know off) that will allow it to be reloaded for an existing user.
